[RISCV] Move performCombineVMergeAndVOps into RISCVFoldMasks

lukel97 opened the pull request:
Continuing on from 72e6c1c70d5e07bbc8cb7cae2ed915108daf93aa, this moves the
vmerge fold peephole into RISCVFoldMasks.

Instead of reasoning in chain and glue nodes, this sinks the True operand to
just before the vmerge pseudo so it can reuse its VL/mask/false/passthru
operands. Most of the time we're able to do this without any hitches because we
check that the vmerge is the only user of True. However there are certain
cases where we have write-after-read dependencies of physical registers that we
also need to sink too, e.g. fcsr in @vmerge_vfcvt_rm in
test/CodeGen/RISCV/rvv/rvv-peephole-vmerge-masked-vops.ll

We also need to implement the masked -> unmasked peephole, since the vmerge
fold peephole can produce new masked pseudos with all one masks that should
then be unmasked. However that peephole is still kept in RISCVISelDAGToDAG to
minimize the diff for this patch. It will be switched over in a later patch.

lukel97 committed:
[RISCV] Move performCombineVMergeAndVOps into RISCVFoldMasks

Continuing on from 72e6c1c70d5e07bbc8cb7cae2ed915108daf93aa, this moves the
vmerge fold peephole into RISCVFoldMasks.

Instead of reasoning in chain and glue nodes, this sinks the True operand to
just before the vmerge pseudo so it can reuse its VL/mask/false/passthru
operands. Most of the time we're able to do this without any hitches because we
check that the vmerge is the only user of True. However there are certain
cases where we have write-after-read dependencies of physical registers that we
also need to sink too, e.g. fcsr in @vmerge_vfcvt_rm in
test/CodeGen/RISCV/rvv/rvv-peephole-vmerge-masked-vops.ll

We also need to implement the masked -> unmasked peephole, since the vmerge
fold peephole can produce new masked pseudos with all one masks that should
then be unmasked. However that peephole is still kept in RISCVISelDAGToDAG to
minimize the diff for this patch. It will be switched over in a later patch.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -88,6 +107,313 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskCopy) {
+}
+
+/// Inserts an operand at Idx in MI, pushing back any operands.
+static void insertOperand(MachineInstr &MI, MachineOperand MO, unsigned Idx) {
```
`MachineInstr::insert` that was added in #67699 may help.

lukel97 replied:
Oh that's right, I saw the patch but forgot to come back to it. Thanks for reminding me

lukel97 commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -9,15 +9,29 @@
 //===---------------------------------------------------------------------===//
 
 #include "RISCV.h"
+#include "RISCVISelDAGToDAG.h"
```
We need RISCVISelDAGToDAG.h since that's where the RISCVMaskedPseudos table is defined, and it's still used by doPeepholeMaskedRVC. The next and final patch in this series would move the table to over to RISCVFoldMasks.cpp

lukel97 committed:
Keep a more canonical form by copying masks from a virtual register first

lukel97 committed:
Remove unnecessary includes

lukel97 committed:
Use MachineInstr::insert

preames commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -88,6 +98,304 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskCopy) {
+#endif
+
+  // Sink True down to MI so that it can access MI's operands.
+  if (!sinkInstructionAndDeps(TrueMI, MI))
```
I think you might be making this more complicated than it needs to be.   My mental model of this transform is that we're mutating TrueMI so as to make the merge operation a nop.  I don't think we need to (or want to) move TrueMI's operands at all.  Given the preconditions we check, I believe it is the case that False and MinVL must be operands of TrueMI.  The only remaining operand we *add* is the mask itself.  

For the mask, we have two options.
1) If the definition of V0 dominates TrueMI, then the mask is available.  In some cases, we may be able to move both the definition of the mask vreg and the copy to V0 before TrueMI.  This is a speculation problem.
2) We can sink TrueMI itself to the masked op.  We know that we *don't* need to sink any it's operands.  All we need to prove is that sinking TrueMI is sound (i.e. if TrueMI can fault in an observable way, we can't sink it below a branch or other observable fault).  Note that condition is basically the inverse of proving the mask is safe to move upwards.  (And the dual of the glue reasoning in SDAG.)

That last bit about being the dual of the glue reasoning may also be important.  The SDAG version isn't infinitely powerful, so we may not need to worry about moving around anything which "could" fault just to preserve equal power.

If I was writing this patch, I'd start by writing the simple version which tries to hoist the mask in trivial cases, and see if that was good enough.  

One thing I can't help but think - all of this would be *much* easier without the copies to V0.  Any idea why we have that?  It seems like the pseudo's could have a vreg operand with a register class which only contained V0 and have the same effect.  Why do we need to manually insert the copies so early?  I'm probably missing something, but what?  (@topperc Any context I'm missing here?)

lukel97 replied:
> I don't think we need to (or want to) move TrueMI's operands at all.

I've named this method rather misleadingly: It only sinks write-after-read dependencies of physical registers for reasons I'll explain below, not any of TrueMI's operands.

As for "hoist the mask" versus "sink TrueMI", I settled on the latter since it ended up being the simpler option.
The ordering of instructions from SelectionDAG typically looks like this:

```
%true = PsuedoVADD ...
%mask = PseudoVMSET ...
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...
```

Where the mask is pretty much always between TrueMI and the vmerge because of the glue. So it never ends up dominating TrueMI in practice.

If we hoist the mask, the trivial case is that it's a VMSET with only some immediate operands.
However a lot of the time the mask will be computed and we'll quickly hit the wall with a naïve implementation, since there'll be a bunch of instructions that need hoisted e.g.

```
%true = PsuedoVADD ...
// All of these need hoisted
%m1 = PseudoVFOO ...
...
%mn = PseudoVBAR ...
%mask = PseudoVMSEQ %mn
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...
```

If we sink TrueMI, there's usually much less we need to do: We check beforehand that the vmerge is the only user of TrueMI, so 99.99% of the time we can just call `TrueMI.beforeBefore(MI)`.

For loads/faulting instructions we also just need to check that there's no stores in between and `isSafeToMove` returns true. However, the annoying case is for pseudos that also read a physical register, e.g.:

```
$frm = FSRM foo
%true = PseudoVFCVT_RM_X_F ... implicit $frm
$frm = FSRM bar
%mask = PseudoVMSET ...
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...

// can't just sink %true:

$frm = FSRM foo
$frm = FSRM bar <-- WAR dependency
%mask = PseudoVMSET ...
$v0 = COPY %mask
%true = PseudoVFCVT_RM_X_F ... implicit $frm <-- incorrect $frm
PsuedoVMERGE $v0, %true, ...
```

In which case we need to sink those write-after-read dependencies too. But this is a much rarer case in comparison to hoisting the mask with lots of dependencies, which is why I ended up settling on the sinking approach.

I am in agreement though that I would like to make this simpler though. I just want to clarify how I ended with the sinking approach. 

> One thing I can't help but think - all of this would be much easier without the copies to V0.

That thought crossed my mind as well. Is it because all the vector pseudos have their mask operands defined as `VMaskOp`?:

```tablegen
def VMaskOp : RegisterOperand<VMV0> {
  let ParserMatchClass = VMaskAsmOperand;
  let PrintMethod = "printVMaskReg";
  let EncoderMethod = "getVMaskReg";
  let DecoderMethod = "decodeVMaskReg";
}
```

preames replied:
The FRM thing is something I hadn't considered.  Going to have to think about that one a bit, but I have to admit I'm very surprised to see implicit register operands for FRM this early.

lukel97 replied:
Fantastic, that should simplify things a lot.

preames commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -59,22 +70,21 @@ char RISCVFoldMasks::ID = 0;
-bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskCopy) {
-  if (!MaskCopy)
+bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskDef) {
+  if (!MaskDef)
```
Unrelated rename and code motion, please land separately and rebase.

preames commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -133,6 +438,48 @@ bool RISCVFoldMasks::convertVMergeToVMv(MachineInstr &MI, MachineInstr *V0Def) {
   return true;
 }
 
+bool RISCVFoldMasks::convertToUnmasked(MachineInstr &MI,
```
This transform is much simpler and easy to review.  What happens if we move only this one to start with?  Does it cause real regressions?

lukel97 replied:
Unfortunately convertToUnmasked is the one with the very large test diff. I think it also runs before `doPeepholeMergeVVMFold` so it would affect the output of that too

lukel97 replied:
You can see the test diff here: https://github.com/lukel97/llvm-project/commit/01854599f2cfd84899c37f4791d6c4a0784017e1

preames replied:
I didn't understand this, so I took my own shot at it (stealing your code as a starting point).  https://github.com/llvm/llvm-project/pull/72441 shows the result, and has a fixme which explains the coupling.  We could either land the half split state (my change), or leave it rolled in here.

preames commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -133,6 +438,48 @@ bool RISCVFoldMasks::convertVMergeToVMv(MachineInstr &MI, MachineInstr *V0Def) {
+    return false;
+
+  // TODO: Increment all MaskOpIdxs in tablegen by num of explicit defs?
+  unsigned MaskOpIdx = I->MaskOpIdx + MI.getNumExplicitDefs();
```
Move this definition closer to the use.  The current placement is visually confusing as we immediate check a mask, but it's a *different* mask.

lukel97 replied:
Whoops, that's very confusing.

lukel97 committed:
Simplify sinking after #72540

lukel97 commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -87,6 +97,258 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskDef) {
+  if (TII->get(TrueOpc).hasUnmodeledSideEffects())
+    return false;
+
+  // The vector policy operand may be present for masked intrinsics
```
That comment is also not needed any more, will remove

lukel97 commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -87,6 +97,258 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskDef) {
+              Dest)
+          .add(VL)
+          .add(TrueMI.getOperand(RISCVII::getSEWOpNum(TrueMCID)));
+      BuildMI(*MI.getParent(), TrueMI, MI.getDebugLoc(), TII->get(RISCV::COPY),
```
Yeah I think we should handle this. Off the top of my head, almost every instruction that uses V0 will have a copy to V0 immediately before it because of the glue in SelectionDAG. But I remember trying to add an assert for this and there were some edge cases.

On a separate note, I have a local branch where I was able to remove the V0 copies in SelectionDAG by adjusting the tablegen patterns to use the VMV0 register class instead, e.g.:

```diff
@@ -4235,13 +4235,13 @@ class VPatBinaryMaskTA<string intrinsic_name,
                    (result_type result_reg_class:$merge),
                    (op1_type op1_reg_class:$rs1),
                    (op2_type op2_kind:$rs2),
-                   (mask_type V0),
+                   (mask_type VMV0:$vm),
                    VLOpFrag, (XLenVT timm:$policy))),
                    (!cast<Instruction>(inst#"_MASK")
                    (result_type result_reg_class:$merge),
                    (op1_type op1_reg_class:$rs1),
                    (op2_type op2_kind:$rs2),
-                   (mask_type V0), GPR:$vl, sew, (XLenVT timm:$policy))>;
+                   (mask_type VMV0:$vm), GPR:$vl, sew, (XLenVT timm:$policy))>;
```

This would simplify all of this, but I'll return to it after this PR since it has a large test diff.

lukel97 committed:
Address review comments:

- Instead of erasing the old mask, move it after MI. Even though in practice
  there shouldn't be any other users of V0 due to SelectionDAG emitting a copy
  for every user, this should preserve the existing value of V0 anyway. This is
  safe to do because we know there's no other defs of V0 between MaskDef and MI.
- Remove isOpSameAs. Turns out we don't need to check if False is NoRegister,
  because that implicit_def->noreg peephole only operates on passthru operands.
- Check that TrueMI and MI are in the same block

lukel97 commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -87,6 +97,258 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskDef) {
+          .addReg(Dest);
+    }
+
+    TrueMI.setDesc(MaskedMCID);
```
I gave this a try by building up a `SmallVector<MachineOperand>` and then creating a new `MachineInstr` with `MaskedMCID`, but then ran into the issue that it doesn't preserve any of the flags or tied operands. The only way I can think of preserving these is by operating on the operands in place, or by cloning the instruction entirely. But then we still need to insert an operand for the unmasked vmv.v.v case. Any ideas?

lukel97 commented:
Superseded by #106108 and others
